Adopt more std::span in WebGPU code

cdumez opened the pull request:
#### 1c1b3b86b53bd5c1b729718ef6e988b977338078
<pre>
Adopt more std::span in WebGPU code
<a href="https://bugs.webkit.org/show_bug.cgi?id=274816">https://bugs.webkit.org/show_bug.cgi?id=274816</a>

Reviewed by Mike Wyrzykowski and Darin Adler.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::GPUQueue::writeTexture):
(WebCore::GPUQueue::copyExternalImageToTexture):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.cpp:
(WebCore::WebGPU::QueueImpl::writeTexture):
(WebCore::WebGPU::QueueImpl::writeBufferNoCopy):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUQueue.h:
* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
(wgpuQueueWriteBuffer):
(wgpuQueueWriteTexture):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::updateMinMaxDepths):
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp:
(WebKit::RemoteQueue::writeTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::writeTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:

Canonical link: <a href="https://commits.webkit.org/279456@main">https://commits.webkit.org/279456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df29587befa549a19b0db9f95f38c0f4a543d175

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53550 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56830 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4276 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55856 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/40408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4074 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43407 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2809 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55648 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31100 "Exiting early after 60 failures. 16175 tests run. 7 new passes 60 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46280 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24546 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27948 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3599 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2432 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3771 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58427 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28708 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3786 "Found 1 new test failure: media/track/track-text-track-destructor-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50809 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29914 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/46454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50148 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30843 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7886 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29686 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -733,7 +733,7 @@ ExceptionOr<void> GPUQueue::copyExternalImageToTexture(ScriptExecutionContext& c
             copyDestination.mipLevel = INT_MAX;
 
-        m_backing->writeTexture(copyDestination, newImageBytes ? newImageBytes : imageBytes.data(), sizeInBytes, dataLayout.convertToBacking(), convertToBacking(copySize));
+        m_backing->writeTexture(copyDestination, newImageBytes ? std::span { newImageBytes, sizeInBytes } : imageBytes, dataLayout.convertToBacking(), convertToBacking(copySize));
```
This doesn't look right, mixing span and pointer in the same expression. Surprised it compiles.

cdumez replied:
`imageBytes` is a span. I don't think I am mixing span and pointer?

darinadler replied:
Oops, not sure why I misunderstood.

darinadler commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.cpp:
```diff
@@ -97,13 +96,12 @@ void QueueImpl::writeBufferNoCopy(
 {
     // FIXME: Use checked arithmetic and check the cast
-    wgpuQueueWriteBuffer(m_backing.get(), m_convertToBackingContext->convertToBacking(buffer), bufferOffset, static_cast<uint8_t*>(source.data()) + dataOffset, static_cast<size_t>(size.value_or(source.size() - dataOffset)));
+    wgpuQueueWriteBuffer(m_backing.get(), m_convertToBackingContext->convertToBacking(buffer), bufferOffset, source.subspan(dataOffset, static_cast<size_t>(size.value_or(source.size() - dataOffset))));
```
Can we remove the `static_cast`?

darinadler commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -392,8 +393,8 @@ static void invalidateCommandBuffers(Vector<std::reference_wrapper<CommandBuffer
-    bool noCopy = size >= largeBufferSize;
-    id<MTLBuffer> temporaryBuffer = noCopy ? device->newBufferWithBytesNoCopy(data, static_cast<NSUInteger>(size), MTLResourceStorageModeShared) : device->newBufferWithBytes(data, static_cast<NSUInteger>(size), MTLResourceStorageModeShared);
+    bool noCopy = data.size() >= largeBufferSize;
+    id<MTLBuffer> temporaryBuffer = noCopy ? device->newBufferWithBytesNoCopy(data.data(), static_cast<NSUInteger>(data.size()), MTLResourceStorageModeShared) : device->newBufferWithBytes(data.data(), static_cast<NSUInteger>(data.size()), MTLResourceStorageModeShared);
```
Can we remove these `static_cast`?

darinadler commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -404,7 +405,7 @@ static void invalidateCommandBuffers(Vector<std::reference_wrapper<CommandBuffer
         toBuffer:buffer
         destinationOffset:static_cast<NSUInteger>(bufferOffset)
-        size:static_cast<NSUInteger>(size)];
+        size:static_cast<NSUInteger>(data.size())];
```
Can we remove this `static_cast`?

darinadler commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -698,7 +699,7 @@ static void invalidateCommandBuffers(Vector<std::reference_wrapper<CommandBuffer
         dataByteSize = newData.size();
         bytesPerImage = newBytesPerImage;
-        data = &newData[0];
+        data = { &newData[0], newData.size() };
```
```
data = newData.mutableSpan();
```

cdumez replied:
Oh indeed! Can't believe I missed that.

darinadler commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.cpp:
```diff
@@ -97,13 +96,12 @@ void QueueImpl::writeBufferNoCopy(
     std::optional<Size64> size)
 {
     // FIXME: Use checked arithmetic and check the cast
```
This comment doesn't really make sense any more.

mwyrzykowski reviewed (commented):
Thank you for addressing these - really happy to see memcpySpan :)

mwyrzykowski commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1701,8 +1701,8 @@ WGPU_EXPORT void wgpuQuerySetRelease(WGPUQuerySet querySet) WGPU_FUNCTION_ATTRIB
 WGPU_EXPORT void wgpuQueueSubmit(WGPUQueue queue, size_t commandCount, WGPUCommandBuffer const * commands) WGPU_FUNCTION_ATTRIBUTE;
-WGPU_EXPORT void wgpuQueueWriteBuffer(WGPUQueue queue, WGPUBuffer buffer, uint64_t bufferOffset, void * data, size_t size) WGPU_FUNCTION_ATTRIBUTE;
-WGPU_EXPORT void wgpuQueueWriteTexture(WGPUQueue queue, WGPUImageCopyTexture const * destination, void * data, size_t dataSize, WGPUTextureDataLayout const * dataLayout, WGPUExtent3D const * writeSize) WGPU_FUNCTION_ATTRIBUTE;
+WGPU_EXPORT void wgpuQueueWriteBuffer(WGPUQueue queue, WGPUBuffer buffer, uint64_t bufferOffset, std::span<uint8_t> data) WGPU_FUNCTION_ATTRIBUTE;
```
This is going to cause an SDK verifier check most likely since I don't think `span` is included at the top of the header. Also may need to wrap this in `#ifdef __cplusplus` or alternatively the entire header in `#ifdef __cplusplus`

mwyrzykowski reviewed (changes requested):
This looks great! But can you please fix the WebGPU.h / span include otherwise I think we will fail SDK verifier check as we have in the past.

Otherwise I think it is good to merge

cdumez committed:
Adopt more std::span in WebGPU code
https://bugs.webkit.org/show_bug.cgi?id=274816

Reviewed by Mike Wyrzykowski and Darin Adler.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::GPUQueue::writeTexture):
(WebCore::GPUQueue::copyExternalImageToTexture):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.cpp:
(WebCore::WebGPU::QueueImpl::writeTexture):
(WebCore::WebGPU::QueueImpl::writeBufferNoCopy):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUQueue.h:
* Source/WebGPU/WebGPU/Queue.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
(wgpuQueueWriteBuffer):
(wgpuQueueWriteTexture):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::updateMinMaxDepths):
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp:
(WebKit::RemoteQueue::writeTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::writeTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:

Canonical link: https://commits.webkit.org/279456@main
